README.md
# Mustang Streamer control (v.0.4)

![Mustang Streamer logo](https://www.oluo.it/public/mustang1_1480x320.jpg)

Mustang Streamer is a personal project that runs Volumio3 installed on a Raspberry CM4 with HiFiBerry sound card, 11.9" touch display, led pushbutton (changing color based on audio sample rate) and POE powered. Only two cables: Ethernet and Optical output.

This software works on Raspberry CM4, Raspberry Pi 3, Raspberry Pi 4. Feel free to test on other platforms and please report the issues.

## Main features
- Turn off display after desired idle time
- Wake up display automatically when player resume
- Led changing color with different sample-rates
- Wake up with button
- Play/pause with button short press
- Power off or reboot CM4 with button long press

![Mustang Streamer](https://www.oluo.it/public/mustang_main.jpg)


## RGB Led configuration

### PCM Formats (FLAC, MP3, Tidal, Spotify...)

| Sample rate | 16bit | 24bit |
|--------|--------|--------|
| <44kHz | Red    | -      |
| 44kHz  | Green  | Yellow |
| 48kHz  | Yellow | Blue   |
| 88kHz  | Blue   | Purple |
| 96kHz  | Purple | Cyan   |
| 192kHz | Cyan   | White  |

### DSD/DSF Formats

| Format  | 1bit   |
|---------|--------|
| DSD64   | Yellow |
| DSD128  | Blue   |
| DSD256  | Purple |
| DSD512  | Cyan   |
| DSD1024 | White  |


___

# Hardware
- [Waveshare POE IO Board for Raspberry CM4](https://www.waveshare.com/product/compute-module-4-poe-board-b.htm)
- Raspberry CM4 8Gb emmc
- [Waveshare 11.9 inch touch display](https://www.waveshare.com/11.9inch-hdmi-lcd.htm)
- [Pushbutton with RGB led](https://it.rs-online.com/web/p/interruttori-a-pulsante/1759645)
- 3x 120Ohm 1/4W resistors
- 1x 1k 1/4W resistor
- 1x 10k 1/4W resistor
- Raspberry compatible soundcard (in this project is [DIGI2 Pro by HiFiBerry](https://www.hifiberry.com/shop/boards/hifiberry-digi2-pro/))
- Adapter cables (based on your enclosure)
- Some spare GPIO Jumper cables
- RCA Panel connector
- 1x RCA Jack
- Shielded audio cable
- Passthrough Ethernet connector
- Cat 5 Ethernet cable (short)
- Shrinking hose
- Hey, obiouvsly, a box! 



![Internal view](https://www.oluo.it/public/mustang3-int.jpg)


Waveshare wide touch display is very ***very*** nice! 

..but making the square hole in the front panel is a nightmare, this display have about 2mm border and a flatcable on one side.. my suggestion: rasps and files (and be patient, self control, yoga, some cigarettes, no swear)!

## Schematics

![GPIO](https://www.oluo.it/public/mustang_gpio.jpg)

Used pins:

| HW        | PIN BCM |
|-----------|---------|
| Switch    | GPIO17  |
| Led Red   | GPIO23  |
| Led Green | GPIO24  |
| Led Blue  | GPIO25  |

Please note, the RGB led in the switch is ***common anode*** without internal resistor.


# Installation
Make sure you have enabled SSH access to your volumio installation. 

From a terminal, log into your device with `ssh volumio@YOURVOLUMIOIP` and:

```
git clone https://github.com/i0466lt/mustang_streamer.git
cd mustang_streamer
./setup.sh
```
Reboot your device with `sudo reboot`

# Basic configuration

## Volumio plugin
Install ***Now playing*** and ***Touch display*** plugin from Volumio3 plugin menu

## Volumio3 configuration

- Under installed plugins, set ***Screensaver timeout*** to 0 in Touch Display Configuration

- Make your favourite layout by editing parameters in Now Playing Configuration

- If you want, you can set Mustang Streamer background for idle screen by selecting `mustang1_1480x320.jpg` under Now Playing Configuration -> Idle Screen

![Mustang streamer idle background](https://www.oluo.it/public/mustang1_1480x320.jpg)



## OS tweaks

### /boot/userconfig.txt
Add these lines to your /boot/userconfig.txt to set your Waveshare display. (Also enables USB for touchscreen)
```
# Enable USB on Waveshare POE IO Board
dtoverlay=dwc2,dr_mode=host


#### Config for waveshare 11.9" HDMI touch display
hdmi_force_hotplug=1
max_framebuffer_height=1480
hdmi_group=2
hdmi_mode=87
hdmi_timings=320 0 80 16 32 1480 0 16 4 12 0 0 0 60 0 42000000 3
## Display rotation
display_rotate=3 #1: 90; 2: 180; 3: 270
```

## Main configuration file (config.ini)

This file contains the pin number assignments and display timeout (in seconds).

Change settings based on your needs.

| Parameter | default | options |
|-----------|---------|---------|
| ledRed | 23 | Control red part of RGB led |
| ledGreen | 24 | Control green part of RGB led |
| ledBlue | 25 | Control blue part of RGB led |
| button | 17 | Button pin |
| timeout_display | 1200 | Seconds before turn off display |
| long_press | reboot | Choose between poweroff or reboot |



# Update

Login into Volumio device via SSH, enter the `mustang_control` directory and issue `./update.sh`



# HOW-TO: Use other GPIO Pin

You can change the PIN Number in `config.ini` file.

Make sure the desired GPIO Pins are unused. You can check by running `gpio readall` from terminal.


If you don't have the `gpio` command install ***WiringPi unofficial fork*** at [https://github.com/WiringPi/WiringPi](https://github.com/WiringPi/WiringPi). I suggest to install the unofficial fork because WiringPi development ceased some years ago. The unofficial fork works and grant support to new hardware like CM4.


## Note for Raspberry CM4 users

If you are running a Raspberry CM4, make sure you are running `gpio` version ***>=2.70*** otherwise the CM4 GPIO is not recognized by `gpio readall` command. You can check your version by type `gpio -v` in the volumio device terminal.

Example fom my device:

```
gpio version: 2.70
Copyright (c) 2012-2018 Gordon Henderson
This is free software with ABSOLUTELY NO WARRANTY.
For details type: gpio -warranty

Raspberry Pi Details:
  Type: CM4, Revision: 00, Memory: 2048MB, Maker: Sony 
  * Device tree is enabled.
  *--> Raspberry Pi Compute Module 4 Rev 1.0
  * This Raspberry Pi supports user-level GPIO access.
```

# Mustang Streamer logo

Yes, I love Deftones! The horse on the logo comes from the album ***white pony***. I've used as part of this useless/non-commercial/diy/free project. (Chino, if you disagree please drop me a mail!)

If you want to make more background images with this logo, you'll find the SVG vector file under the `logo` folder.


## Background image

Background image is based on artwork by [freepik.com](https://it.freepik.com/foto-gratuito/composizione-astratta-della-luce-ultravioletta-uv_21794910.htm#query=background%20neon&position=9&from_view=search&track=ais)


# TO-DO

- Much to do!!

Feel free to open issues, comment and contribute!



CHANGELOG.md
# Changelog

### Version 0.4
- Fixed gpio and vcgencmd permissions 

### Version 0.3
- Added function to choose action on button long press (poweroff or reboot) 
- Turn off led on service stop or manual abort

### Version 0.2
- Added external config file
- Added background image for idle screen

### Version 0.1
- Initial version 

setup.sh
#!/bin/bash
#
# Setup Mustang Streamer services, libraries, utils and backgrounds

echo -e "Mustang streamer setup.."

# Copy backgrounds to data folder
cp app_setup/*.jpg /data/backgrounds/

# GPIO
sudo pip3 install --upgrade RPi.GPIO


# Python3 modules
sudo pip3 install requests


# WiringPi unofficial fork (Support new devices like CM4)
git clone https://github.com/WiringPi/WiringPi.git
cd WiringPi
./build

# Finishing..
sudo ldconfig

# Adding various permissions
sudo echo "volumio ALL=(ALL) NOPASSWD: /usr/bin/vcgencmd" >> /etc/sudoers
sudo usermod -a -G kmem volumio
sudo chown root.gpio /dev/gpiomem
sudo chmod g+rw /dev/gpiomem




# Setting up services
sudo cp /home/volumio/mustang_control/app_setup/mustang_control.service /lib/systemd/system/
sudo systemctl daemon-reload
sudo systemctl enable mustang_control.service
echo -e "\nStarting Mustang streamer service.."
sudo systemctl start mustang_control.service
sudo systemctl status mustang_control.service

echo -e "\n\nInstallation complete!\n"

mustang_control.py
#!/usr/bin/env python3

###############################
# Mustang streamer controller #
###############################

import requests
import os
import time
import RPi.GPIO as GPIO
import json
import subprocess
import configparser



#### External config file
basedir = '/home/volumio/mustang_control/'
fileConfig = basedir + 'config.ini'
cfg_file = configparser.ConfigParser()
cfg_file.read(fileConfig)

# RGB Led pins (BCM)
pinR = int(cfg_file['mustang streamer config']['ledRed']) # default:23
pinG = int(cfg_file['mustang streamer config']['ledGreen']) # default:24
pinB = int(cfg_file['mustang streamer config']['ledBlue']) # default:25


# Button
# 3.3v -> switch ---> 10k -> GND
#                |_ 1k -> gpio_pin (pinRESET)	
pinRESET = int(cfg_file['mustang streamer config']['button']) # default:17
button_action = str(cfg_file['mustang streamer config']['long_press']) # default:poweroff
LONG_PRESS_TIME = 3.0  # in seconds


# Display poweroff timer
time_off = int(cfg_file['mustang streamer config']['timeout_display']) # Timeout display (in seconds) default:1200sec


### SETUP
GPIO.setmode(GPIO.BCM)
GPIO.setwarnings(False)

# Button functions
def button(channel):

	print("Button pressed")

	start_time = time.time()
	
	while GPIO.input(channel) == GPIO.HIGH:
		# Check if is a long press or normal button press
		if time.time() - start_time > LONG_PRESS_TIME:
			print("- Long press! (more than "+str(LONG_PRESS_TIME)+" seconds)")
			os.system("volumio stop")
			led_Red()
			os.system("sudo systemctl " + button_action )
			return
		else:
			# Code for short press
			os.system('volumio toggle')
			set_timer_display()

		time.sleep(1)



# GPIO: Setup button
GPIO.setup(pinRESET, GPIO.IN, GPIO.PUD_DOWN)
GPIO.add_event_detect(pinRESET, GPIO.RISING, callback=button, bouncetime=4000)


# GPIO: Setup RGB led pins
GPIO.setup(pinR,GPIO.OUT)
GPIO.setup(pinG,GPIO.OUT)
GPIO.setup(pinB,GPIO.OUT)

### General config
quality_txt = "-"		# Placeholder
rainbow_timer = 0.35	# Startup led rainbow wait time


##### FUNCTIONS #####
# Volumio status
def status_volumio():
	# Volumio status
	response = requests.get("http://localhost:3000/api/v1/getState")
	volumio = response.json()

	if "samplerate" in volumio:
		raw_samplerate = str(volumio["samplerate"]).split()
		raw_bit = str(volumio["bitdepth"]).split()
		audio_samplerate = float(raw_samplerate[0])
		audio_bitrate = int(raw_bit[0])
	else:
		audio_samplerate = 0
		audio_bitrate = 0

	if "status" in volumio:	

		status = volumio["status"]

		if status == "play":
			print("In riproduzione")
			# Reset display sleep timer
			display_poweron()
			reset_timer_display()

			### AUDIO QUALITY ###
			# Audio DSD/DSF
			if audio_bitrate == 1: 

				if audio_samplerate < 5.6:
					quality_txt = "DSD64"
					led_Yellow()

				elif audio_samplerate < 11.2:
					quality_txt = "DSD128"
					led_Blue()

				elif audio_samplerate < 22.5:
					quality_txt = "DSD256"
					led_Purple()

				elif audio_samplerate < 45.1:
					quality_txt = "DSD512"
					led_Cyan()

				elif audio_samplerate >= 45.1:
					quality_txt = "DSD1024"
					led_White()

				else:
					quality_txt = "DSD format: unknown!"
					led_Red()

			# Audio 16bit
			elif audio_bitrate == 16:
				# Standard bitrate

				if audio_samplerate < 44:
					# Low quality
					quality_txt = "LOW"
					led_Red()

				elif 44 < audio_samplerate < 48:
					# 16bit - 44kHz
					quality_txt = "CD"
					led_Green()

				elif 48 <= audio_samplerate < 88:
					# 16bit - 48kHz
					quality_txt = "SACD"
					led_Yellow()

				elif 88 <= audio_samplerate < 96:
					# 16bit - 88kHz
					quality_txt = ""
					led_Blue()

				elif 96 <= audio_samplerate < 192:
					# 16bit - 96kHz
					quality_txt = ""
					led_Purple()

				elif audio_samplerate == 192:
					# 16bit - 192kHz
					quality_txt = ""
					led_Cyan()

				else: 
					# Unknown sample rate
					quality_txt = "Sample rate error"
					print("### Undefined: "+volumio["samplerate"]+" "+volumio["bitdepth"])


			# Audio 24bit
			elif audio_bitrate == 24:
				# High bitrate

				if 44 < audio_samplerate < 48:
					# 24bit - 44kHz
					quality_txt = ""
					led_Yellow()

				elif 48 <= audio_samplerate < 88:
					# 24bit - 48kHz
					quality_txt = ""
					led_Blue()

				elif 88 <= audio_samplerate < 96:
					# 24bit - 88kHz
					quality_txt = ""
					led_Purple()

				elif 96 <= audio_samplerate < 192:
					# 24bit - 96kHz
					quality_txt = ""
					led_Cyan()

				elif audio_samplerate == 192:
					# 24bit - 192kHz
					quality_txt = ""
					led_White()

				else: 
					# Unknown sample rate
					quality_txt = "Sample rate error"
					print("### Undefined: "+volumio["samplerate"]+" "+volumio["bitdepth"])
					led_Off()


			elif audio_bitrate == 0:
				# Transaction between local and streaming service
				quality_txt = "Switching music service..."
				led_Off()


			# Unknown format
			else:
				quality_txt = "Unknown format"
				led_Off()



				
			# Print audio quality on terminal
			print("Audio quality: "+quality_txt+" "+str(volumio["bitdepth"])+" "+str(volumio["samplerate"]) )


		elif status == "pause":
			print("Paused")
			led_Off()
			set_timer_display()
		else:
			print("Stopped")
			led_Off()
			set_timer_display()

	else:
		# Unknown status
		print("### Status not found! ###")
		led_Off()


# Volumio play/pause 
def volumio_playpausa():
	checkplay = requests.get("http://localhost:3000/api/v1/getState")
	playstatus = checkplay.json()

	if "status" in playstatus:
		stato = playstatus["status"]

		if stato == "play":
			os.system('volumio pause')
			set_timer_display()
		
		elif (stato == "pause" or stato == "stop"):
			display_poweron()
			os.system('volumio play')
			reset_timer_display()
		else:
			print("Unknown player status!")

	else:
		print("No status! Is Volumio running?")


# Display power off
def display_poweroff():
	output = subprocess.getoutput("sudo vcgencmd display_power")
	rawOut = output.split('=')
	verifica = int(rawOut[1])
	if verifica > 0:
		print('Turning off display!')
		os.system('sudo vcgencmd display_power 0')
		reset_timer_display()


# Display power on
def display_poweron():
	output = subprocess.getoutput("sudo vcgencmd display_power")
	rawOut = output.split('=')
	verifica = int(rawOut[1])
	if verifica == 0:
		print('WAKEUP display!')
		os.system('sudo vcgencmd display_power 1')


# Power off display at sleep timer
def sleep_display():
	rawTimer = open("/tmp/timer_display.dat", "r")
	for line in rawTimer.readlines():
		last_timer = int(line)
	rawTimer.close()

	if last_timer > 0:
		tsNow = int( time.time() )
		# Found a timestamp
		ts_limite = last_timer + time_off
		test = ts_limite - tsNow

		if tsNow > ts_limite:
			display_poweroff()


# Reset display sleep timer
def reset_timer_display():
	rawTimer = open("/tmp/timer_display.dat", "r")
	for line in rawTimer.readlines():
		last_timer = int(line)
	rawTimer.close()
	if last_timer > 0:
		timerfile = open("/tmp/timer_display.dat", "w")
		timerfile.write("0")
		timerfile.close()


# Set display sleep timer
def set_timer_display():
	rawTimer = open("/tmp/timer_display.dat", "r")
	for line in rawTimer.readlines():
		last_timer = int(line)
	rawTimer.close()
	if last_timer == 0:
		print('- Setting up display sleep timer')
		tsNow = str ( int( time.time() ) )
		timerfile = open("/tmp/timer_display.dat", "w")
		timerfile.write(tsNow)
		timerfile.close()




## RGB Led definitions
def led_Off():
	GPIO.output(pinR,GPIO.HIGH)
	GPIO.output(pinG,GPIO.HIGH)
	GPIO.output(pinB,GPIO.HIGH)

def led_Green():
    GPIO.output(pinR,GPIO.HIGH)
    GPIO.output(pinG,GPIO.LOW)
    GPIO.output(pinB,GPIO.HIGH)

def led_Red():
    GPIO.output(pinR,GPIO.LOW)
    GPIO.output(pinG,GPIO.HIGH)
    GPIO.output(pinB,GPIO.HIGH)

def led_Blue():
    GPIO.output(pinR,GPIO.HIGH)
    GPIO.output(pinG,GPIO.HIGH)
    GPIO.output(pinB,GPIO.LOW)

def led_White():
    GPIO.output(pinR,GPIO.LOW)
    GPIO.output(pinG,GPIO.LOW)
    GPIO.output(pinB,GPIO.LOW)

def led_Yellow():
    GPIO.output(pinR,GPIO.LOW)
    GPIO.output(pinG,GPIO.LOW)
    GPIO.output(pinB,GPIO.HIGH)

def led_Purple():
    GPIO.output(pinR,GPIO.LOW)
    GPIO.output(pinG,GPIO.HIGH)
    GPIO.output(pinB,GPIO.LOW)

def led_Cyan():
    GPIO.output(pinR,GPIO.HIGH)
    GPIO.output(pinG,GPIO.LOW)
    GPIO.output(pinB,GPIO.LOW)

# Rainbow
def led_rainbow(quanti):
	for i in range(quanti):	
		print("Rainbow!")
		led_White()
		time.sleep(rainbow_timer)
		led_Purple()
		time.sleep(rainbow_timer)
		led_Blue()
		time.sleep(rainbow_timer)
		led_Cyan()
		time.sleep(rainbow_timer)
		led_Green()
		time.sleep(rainbow_timer)
		led_Yellow()
		time.sleep(rainbow_timer)
		led_Red()
		time.sleep(rainbow_timer)
	led_Off()

###### END FUNCTIONS #######



### INIT ###
led_Off()
print("##### Mustang Streamer starting.. #####\n")
led_rainbow(1)

# Set first display timer
timerfile = open("/tmp/timer_display.dat", "w")
timerfile.write("0")
timerfile.close()



###### Main loop
try:
	while True:

		status_volumio()

		sleep_display()

		time.sleep(0.3)

except KeyboardInterrupt:
	print("\nUser interrupt..")
	led_Off()
	GPIO.cleanup()

## Last
led_Off()
GPIO.cleanup()
